libraries/discriminator/src/lib.rs

```diff
@@ -21,34 +21,81 @@ mod tests {
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("some_discriminator_hash_input")]
+    #[discriminator_hash_input("my_first_instruction")]
     pub struct MyInstruction1 {
         arg1: String,
         arg2: u8,
     }
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("yet_another_discriminator_hash_input")]
-    pub struct MyInstruction2 {
-        arg1: u64,
+    #[discriminator_hash_input("global:my_second_instruction")]
+    pub enum MyInstruction2 {
+        One,
+        Two,
+        Three,
     }
 
     #[allow(dead_code)]
     #[derive(SplDiscriminate)]
-    #[discriminator_hash_input("global:my_instruction_3")]
-    pub enum MyInstruction3 {
-        One,
-        Two,
-        Three,
+    #[discriminator_hash_input("global:my_instruction_with_lifetime")]
+    pub struct MyInstruction3<'a> {
+        data: &'a [u8],
+    }
+
+    #[allow(dead_code)]
+    #[derive(SplDiscriminate)]
+    #[discriminator_hash_input("global:my_instruction_with_one_generic")]
+    pub struct MyInstruction4<T> {
```

Comment:
Is there a particular use case for a generic? I worry that it could cause more confusion than anything else. If I'm defining an instruction or struct in an interface, I probably want the type to be well-defined, otherwise people could deserialize or interpret it incorrectly.

On the flipside, since it isn't a huge lift to add, I'm thinking we can include it and see how people deal with it.

Reply:
Yeah, I felt like it was more important to cover our bases where someone might need a discriminator for some use case where their struct has a generic field, and all of a sudden the macro fails bc they have the arg.
Keep in mind - even though the discriminators we're using here are going to be primarily for interfaces - we are trying to de-couple this lib from just that particular use case

Reply:
It kinda goes against the point of a discriminator though. A discriminator is meant to be a unique tag for a type. With the ability to define it on a generic type, `MyInstruction<u8>` and `MyInstruction<String>` have the same discriminator, when they are different concrete types. What can any code do with that? But there's probably some other use case that I'm missing.

Reply:
So you think by default we shouldn't support generic types, then? Only lifetimes?
One use case that comes to mind for me is something like a TLV type with one generic field.

Maybe you're packing a buffer with TLV types and want to be able to flop between a few different types of data for one field, ie:

```
MyTlvData<T> {
   name: String,
   data: T,
}
```

Imagine even moreso if the type `T` is _another_ TLV struct (and can flop between different ones for nested TLV)

I see your point, but in my head the above struct is still the same data, especially if it's trait-scoped pretty well.

Ultimately my thought was leave it up to the user to decide if a generic should be used based on your uniqueness arguments

Reply:
Maybe I'm thinking they shouldn't have unique discriminators and sounds like you think they def should (I'm not sure who is right lol).

It sounds like for the following example - where I'm going to create a slab of bytes with N number of TLV entries for type `Vault` - you'd rather someone use different types altogether (ie. different discriminators)

```
#[spl_discrim("one_vaulted_token")]
struct Vault<V> where V: SplToken {
   authority: Pubkey,
   token: V,
}

#[derive(SplToken)]
struct Usdc {}

#[derive(SplToken)]
struct Bonk {}

...
let first_tlv_entry = Usdc::new();
let second_tlv_entry = Bonk::new();
tlv_state.pack(first_tlv_entry);
tlv_state.pack(second_tlv_entry);
...
```

You could also go with nested TLV:

```
#[spl_discrim("one_vaulted_token")]
struct Vault<V> where V: SplDiscriminate {
   authority: Pubkey,
   token: V,
}

#[spl_discrim("vaulted_usdc")]
struct Usdc {}

#[spl_discrim("vaulted_bonk")]
struct Bonk {}

...
let first_tlv_entry = Usdc::new();
let second_tlv_entry = Bonk::new();
tlv_state.pack(first_tlv_entry);
tlv_state.pack(second_tlv_entry);
...
```

Reply:
Yeah I guess someone could do this. The crucial bit to note here is that these examples don't work unless we allow for multiple entries of the same type.